service/worker/sysworkflow/system_workflow.go

```diff
@@ -22,101 +22,337 @@ package sysworkflow
 func SystemWorkflow(ctx workflow.Context) error {
-	id := workflow.GetInfo(ctx).WorkflowExecution.ID
-	logger := workflow.GetLogger(ctx)
-	scope := workflow.GetMetricsScope(ctx).Tagged(map[string]string{SystemWorkflowIDTag: id})
-	ch := workflow.GetSignalChannel(ctx, SignalName)
-
-	logger.Info("started new system workflow")
+	sysWorkflowInfo := workflow.GetInfo(ctx)
+	logger := globalLogger.WithFields(bark.Fields{
+		logging.TagWorkflowExecutionID: sysWorkflowInfo.WorkflowExecution.ID,
+		logging.TagWorkflowRunID:       sysWorkflowInfo.WorkflowExecution.RunID,
+	})
+	logger.Info("started system workflow")
+	globalMetricsClient.IncCounter(metrics.SystemWorkflowScope, metrics.SysWorkerWorkflowStarted)
+	sw := globalMetricsClient.StartTimer(metrics.SystemWorkflowScope, metrics.SysWorkerContinueAsNewLatency)
+	ch := workflow.GetSignalChannel(ctx, signalName)
 	signalsHandled := 0
-	for ; signalsHandled < SignalsUntilContinueAsNew; signalsHandled++ {
+	for ; signalsHandled < signalsUntilContinueAsNew; signalsHandled++ {
 		var signal signal
 		if more := ch.Receive(ctx, &signal); !more {
-			scope.Counter(ChannelClosedUnexpectedlyError).Inc(1)
-			logger.Error("cadence channel was unexpectedly closed")
 			break
 		}
-		selectSystemTask(scope, signal, ctx, logger)
+		globalMetricsClient.IncCounter(metrics.SystemWorkflowScope, metrics.SysWorkerReceivedSignal)
+		selectSystemTask(signal, ctx, logger, globalMetricsClient)
 	}
 
 	for {
 		var signal signal
 		if ok := ch.ReceiveAsync(&signal); !ok {
 			break
 		}
-		selectSystemTask(scope, signal, ctx, logger)
+		globalMetricsClient.IncCounter(metrics.SystemWorkflowScope, metrics.SysWorkerReceivedSignal)
+		selectSystemTask(signal, ctx, logger, globalMetricsClient)
 		signalsHandled++
 	}
-
-	logger.Info("completed current set of iterations, continuing as new",
-		zap.Int(logging.TagIterationsUntilContinueAsNew, signalsHandled))
-
-	ctx = workflow.WithExecutionStartToCloseTimeout(ctx, WorkflowStartToCloseTimeout)
-	ctx = workflow.WithWorkflowTaskStartToCloseTimeout(ctx, DecisionTaskStartToCloseTimeout)
-	return workflow.NewContinueAsNewError(ctx, SystemWorkflowFnName)
+	ctx = workflow.WithExecutionStartToCloseTimeout(ctx, workflowStartToCloseTimeout)
+	ctx = workflow.WithWorkflowTaskStartToCloseTimeout(ctx, decisionTaskStartToCloseTimeout)
+	globalMetricsClient.IncCounter(metrics.SystemWorkflowScope, metrics.SysWorkerContinueAsNew)
+	sw.Stop()
+	logger.WithFields(bark.Fields{
+		logging.TagNumberOfSignalsUntilContinueAsNew: signalsHandled,
+	}).Info("system workflow is continuing as new")
+	return workflow.NewContinueAsNewError(ctx, systemWorkflowFnName)
 }
 
-func selectSystemTask(scope tally.Scope, signal signal, ctx workflow.Context, logger *zap.Logger) {
-	scope.Counter(HandledSignalCount).Inc(1)
-
+func selectSystemTask(signal signal, ctx workflow.Context, logger bark.Logger, metricsClient metrics.Client) {
 	ao := workflow.ActivityOptions{
 		ScheduleToStartTimeout: time.Minute,
```

Comment:
Is this the timeout for archival of entire workflow execution history? Do you think 1 minute is good enough for all use cases?  I think currently we are allowing histories to grow as big as 200MB.

Reply:
Good catch. I increased this to 5 minutes to be on the safe side.